### PR TITLE
fix(obsidian-utils): cwd-gone session-id resolution (issue #105)

### LINKED ISSUE #105: _get_session_id_fast() raises on deleted cwd; insight savers fall back to source_session: unknown after mid-session worktree deletion
## Context

Surfaced 2026-04-25 during deep validation of the 3 medium-confidence (conf=0.4–0.6) findings from `/vault-doctor --check source-sessions`. Two of the three cases (`2026-04-24-retro-e2e0.md` and `2026-04-24-retro-4b24.md`) carry `source_session: unknown` despite being saved during real, long-running worktree sessions whose JSONLs still exist on disk.

The mechanism is now well-understood and reproducible.

## Reproduction

```python
import os
# Simulate cwd being a deleted dir (which happens after `gh pr merge --delete-branch`
# from inside the worktree being deleted — the harness's tracked cwd points to a
# now-missing path)
orig = os.getcwd
os.getcwd = lambda: (_ for _ in ()).throw(FileNotFoundError(2, 'No such file or directory'))
try:
    from obsidian_utils import _get_session_id_fast
    sid = _get_session_id_fast()
    print('SID:', repr(sid))
except Exception as e:
    print('EXCEPTION:', type(e).__name__, e)  # → FileNotFoundError
finally:
    os.getcwd = orig
```

Result: `_get_session_id_fast()` raises `FileNotFoundError`. The exception propagates through the python3 -c invocation in every insight skill's session-context helper. The Bash command exits non-zero. The model can't parse `SID=…` from empty stdout. Per SKILL contract (compress, decide, error-log, retro):

> **Important:** If `SESSION_ID` is `unknown`, use `unknown` for `source_session` and omit `source_session_note` entirely.

The model stamps `source_session: unknown` and omits the basename — exactly what we see in the two retros.

## Empirical evidence

Two retros saved 2026-04-24 ~04:05 local time, with `source_session: unknown` and empty `source_session_note`. Content-grep against the 3 active worktree JSONLs from that day proved:
- `2026-04-24-retro-e2e0.md` → originated in issue-50 worktree session `7c71d4da-…` (hash b8a0). Tokens like `subagent-driven 7-task` and `verify-staged-files` appear 4× in issue-50's JSONL, 0× in others.
- `2026-04-24-retro-4b24.md` → originated in issue-68 worktree session `4db2eef6-…` (hash 543f). Tokens `_write_session` (26×), `_short_session_hash` (49×), `Ready to merge` (7×), `BEHIND develop` (3×) all isolated to issue-68's JSONL.

Both retros document a mid-session `gh pr merge --delete-branch` from inside the worktree:

> "the merge deleted the worktree directory, the harness's cwd anchor got wedged to the missing path, and every subsequent Bash call fails with `Path … does not exist` before the command even starts"

After that point, every helper-via-Bash call fails before getting to the python3 step.

Slow-path simulation confirms the helper *would* have produced the right answer if it could run: from cwd `obsidian-brain--issue-50-…` the slow-path glob returns `7c71d4da-…` (the right SID); from `obsidian-brain--issue-68-…` it returns `4db2eef6-…`. So this is not a slow-path defect — it's an `os.getcwd()` precondition that fails when the cwd directory is deleted.

Compounding factor: **no bootstrap files exist** under `~/.claude/obsidian-brain/bootstrap-*` for any of these cwds. The fast-path bootstrap fallback (which would have cached a valid SID before the cwd was deleted) didn't kick in because nothing was ever written there. That's a separate bug (SessionStart should be the authoritative bootstrap writer per the docstring) but compounds this one.

### Primary: harden `_get_session_id_fast()` and `_slow_path_newest_sid()` against gone-cwd

Wrap the `os.getcwd()` call and provide a fallback chain:

```python
def _resolve_project_basename() -> str | None:
    \"\"\"Return cwd basename, falling back to CLAUDE_PROJECT_DIR when cwd is deleted.\"\"\"
    try:
        return os.path.basename(os.getcwd())
    except OSError:
        env = os.environ.get(\"CLAUDE_PROJECT_DIR\")
        if env:
            return os.path.basename(env)
        return None  # caller should treat as 'cannot determine'
```

In `_get_session_id_fast()` and `_slow_path_newest_sid()`:
- Use `_resolve_project_basename()` instead of inline `os.path.basename(os.getcwd())`
- If it returns `None`, fall through to a final \"newest JSONL across all `~/.claude/projects/*/` dirs\" strategy. Less precise but never raises and never returns 'unknown' when ANY JSONL exists. Document the precision trade-off (multi-project ambiguity) in the docstring; the alternative is to keep raising — but that's worse than the current 'unknown' fallback that loses metadata silently.

### Secondary: ensure SessionStart writes the bootstrap

The docstring for `_get_session_id_fast()` says \"The SessionStart hook is the sole authoritative writer of the bootstrap.\" Today there are no bootstrap files for any active or recently-active cwd. SessionStart hook is either failing silently or not writing them at all. Investigate and fix; with a bootstrap file present, the fast-path returns a valid SID even after cwd is deleted (the bootstrap path is keyed by cwd basename which we computed from the env-var fallback above).

### Tertiary: SKILL helper bash prelude should handle `cd` failure explicitly

`cd \"\$(git rev-parse --show-toplevel 2>/dev/null || pwd)\"` silently no-ops when both fail; the rest of the script then runs from whatever the wedged cwd was. Better:

```bash
TARGET=\"\$(git rev-parse --show-toplevel 2>/dev/null || pwd 2>/dev/null || echo \"\$CLAUDE_PROJECT_DIR\")\"
cd \"\$TARGET\" 2>/dev/null || cd \"\$HOME\"
```

Last-resort `cd \$HOME` keeps cwd valid so downstream Python can run.

## Test plan

- Unit test: `_get_session_id_fast()` with monkeypatched `os.getcwd` → `FileNotFoundError`. Assert returns a sensible value (not raises) when CLAUDE_PROJECT_DIR is set.
- Unit test: same, but no CLAUDE_PROJECT_DIR. Assert returns 'unknown' (graceful, not raises).
- Integration test: synth vault + delete cwd between SessionStart and helper invocation. Assert helper still resolves the right SID via bootstrap or env-var fallback.
- Regression: existing happy-path tests still pass.

## Migration / backward compatibility

Pure additive resilience. No behaviour change when cwd is valid. Affects only the failure mode that today produces 'unknown'.

## Risk

Low. The fallback chain is conservative — only fires when the existing path raises or returns nothing.

## Triage

**P2-medium** — silent metadata loss on retros/insights saved after worktree deletion. Two retros affected in this audit; rate scales with how often `gh pr merge --delete-branch` is run from inside the worktree. The `feedback_*_worktree_*` memories already advise against this pattern, but the pattern keeps recurring (last seen 2026-04-24 in both retros' \"What Didn't Work\" sections).

## Related

- #100 — SessionEnd silently drops above-threshold session notes. Compounds: in this audit, issue-45 and issue-50 worktree sessions (1590 / 1915 lines, both long-running) have JSONLs but no vault session notes. issue-68 has both. The retros pointing at unknown sessions also can't be auto-repaired by vault-doctor because there's nothing to point at.
- #101 — write-side basename divergence (separate bug class; this issue is about session-id resolution failure, not basename composition)
- #91 — `/finish: gh pr merge --delete-branch fails local cleanup when target is in another worktree`. Operational-discipline sibling.
- Memory `technical_worktree_cwd_bash_lockup` — describes the wedged-cwd post-deletion mechanism that triggers this
- Memory `feedback_subagent_staged_revert_trap`, `feedback_default_to_feature_branch_not_worktree` — adjacent advisory rules
- Vault insight: `claude-insights/2026-04-25-source-session-note-basename-staleness-root-cause-a701.md` (Update section will be extended with these findings)

### PR DESCRIPTION
**Pull request by abhattacherjee**
## Summary

Phase 1 vault-doctor cluster prevention fix. Adds a layered fallback chain to session-ID resolution in `obsidian_utils.py` so insight savers no longer stamp `source_session: unknown` after mid-session worktree deletion.

**Closes #105**

## Why

When a CC worktree is removed mid-session via `gh pr merge --delete-branch` from inside the worktree, the harness's tracked cwd points at a now-missing directory. Insight-saver SKILLs (`/retro`, `/compress`, `/decide`, `/error-log`) all spawn a `python3 -c '... obsidian_utils ...'` helper whose `os.getcwd()` raises `FileNotFoundError` before resolution logic runs. The SKILL falls back to `SESSION_ID=unknown` and stamps `source_session: unknown` on saved insights — silent metadata loss. Two retros on 2026-04-24 (`2026-04-24-retro-e2e0.md`, `2026-04-24-retro-4b24.md`) carry this exact symptom.

## Design

- Spec: [`docs/superpowers/specs/2026-04-25-issue-105-cwd-gone-session-id-resolution-design.md`](https://github.com/abhattacherjee/spec-docs/blob/spec/issue-105-cwd-gone-session-id-resolution/superpowers/specs/2026-04-25-issue-105-cwd-gone-session-id-resolution-design.md)
- Plan: [`docs/superpowers/plans/2026-04-25-issue-105-cwd-gone-session-id-resolution-plan.md`](https://github.com/abhattacherjee/spec-docs/blob/spec/issue-105-cwd-gone-session-id-resolution/superpowers/plans/2026-04-25-issue-105-cwd-gone-session-id-resolution-plan.md)

## What changed

**Resolver chain (`hooks/obsidian_utils.py`):**

A new private resolver `_resolve_session_id(allow_bootstrap=True)` is the single source of truth for current-session SID resolution. Both `_get_session_id_fast()` and `_slow_path_newest_sid()` collapse to thin wrappers preserving names, signatures, and 16+ existing call sites. Resolution layers:

1. Project basename via `_resolve_project_basename()` — cwd → `CLAUDE_PROJECT_DIR` → `None`
2. Bootstrap fast path (skipped if `allow_bootstrap=False`)
3. Slow-path JSONL glob
4. **NEW** — Recent-bootstrap best-effort scan (skipped if `allow_bootstrap=False`)
5. `"unknown"` sentinel

Layer 4 is gated on the same flag as layer 2 because both are bootstrap-reading mechanisms, and the health-check caller `check_hook_status()` requires bootstrap-blindness to detect SessionStart hook failures.

**Param naming note:** spec used `allow_bootstrap_cache`; implementation uses shorter `allow_bootstrap` (less redundant — `_recent_bootstrap_sid` IS bootstrap-cache reading too).

## Test plan

- [x] **806 → 807 pytest passing** (793 baseline + 14 new tests in `tests/test_get_session_context.py`)
  - 3 unit tests for `_resolve_project_basename` (happy/env-fallback/both-fail)
  - 6 unit tests for `_recent_bootstrap_sid` (zero/one/two/tmp-skip/stale/empty-content)
  - 5 integration tests for `_resolve_session_id` (cwd-gone+layer-4, cwd-gone+unknown, happy-path, slow-path-skips-layer-2, slow-path-skips-layer-4)
- [x] **Fixture script `scripts/dev-test/test-issue-105-manual.sh`** — 18 assertions across 5 phases (function-level + real OS-deletion + bash-prelude integration). Will be 18/18 after `/dev-test install` of this branch.
- [ ] **Live-CC walkthrough** `scripts/dev-test/DEV-TEST-ISSUE-105.md` in sibling session (gate for `/finish`)

## Cluster context

Phase 1 of the vault-doctor source-attribution cluster (parallel with #100, #94). Sibling #101 shipped 2026-04-25 as `6bdd120`.

## Commits

8 small commits in TDD red→green→refactor pattern:

- `ae0b8bd` feat: add `_resolve_project_basename`
- `5872e41` feat: add `_recent_bootstrap_sid`
- `86499b5` refactor: collapse session-id resolution into `_resolve_session_id`
- `3847492` refactor: gate layer 4 on `allow_bootstrap` (preserves bootstrap-blind contract)
- `c8bcf49` test(dev-test): fixture script Phases A+B
- `9d057a3` test(dev-test): fixture script Phases C+D+E
- `8e21318` docs(dev-test): live-CC checklist
- `bb40754` docs(changelog): note fix under [Unreleased]/Fixed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: hooks/obsidian_utils.py [hunk 1]
```diff
@@ -277,6 +277,72 @@ def _safe_getcwd() -> str:
...
+                content = f.read().strip()
+        except OSError:
+            continue
+        if not content:
+            continue
+        candidates.append(content)
+        if len(candidates) > 1:
+            return None  # short-circuit — strict exactly-one
+
+    return candidates[0] if len(candidates) == 1 else None
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply hooks/obsidian_utils.py:358]: ✅ Addressed in d8ab2a3.
**What was done:** Added `_SID_FILENAME_SAFE.fullmatch(content)` validation before appending to candidates. Mirrors the existing validator used at `obsidian_utils.py:76`. Reject path-traversal (`../foo`), absolute paths (`/abs/path`), whitespace, control chars, glob chars, path separators, and length-overflow (>128 chars).
**Regression test:** `test_recent_bootstrap_sid_rejects_unsafe_sid_format` exercises 9 attacker-controlled / corrupted SID shapes; all return None per spec. 808/807 pytest pass.

File: scripts/dev-test/test-issue-105-manual.sh [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+CACHE_DIR=$(find ~/.claude/plugins/cache -type d -path "*/obsidian-brain/*" \
+    -not -path "*.bak*" 2>/dev/null | sort -V | tail -1 | xargs dirname 2>/dev/null || true)
+if [ -z "$CACHE_DIR" ] || [ ! -d "$CACHE_DIR" ]; then
+    echo "❌ Could not locate obsidian-brain plugin cache."
+    echo "   Run /dev-test install first."
+    exit 1
+fi
+
+HOOK_DIR=$(find "$CACHE_DIR" -maxdepth 2 -type d -name hooks | sort -V | tail -1)
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply scripts/dev-test/test-issue-105-manual.sh:32]: ✅ Addressed in d8ab2a3.
**What was done:** Added the suggested HOOK_DIR validity guard verbatim, plus a follow-up precheck that imports `obsidian_utils` and asserts `_resolve_project_basename` exists. This converts the previous "15 silent AttributeErrors with empty got values" pre-install failure mode into an actionable "Plugin cache is missing issue #105 functions. Run /dev-test install..." message. Caught precisely the operator-confusion failure mode you described.

File: hooks/obsidian_utils.py [hunk 3]
```diff
@@ -277,6 +277,78 @@ def _safe_getcwd() -> str:
...
+         fall through to project-agnostic fallbacks
+
+    Never raises. Mirrors the env-fallback contract that obsidian_session_hint.py
+    line 117 already uses for hook input (cwd ← hook payload, getcwd second).
+    """
+    try:
+        return os.path.basename(os.getcwd())
+    except OSError:
+        env = os.environ.get("CLAUDE_PROJECT_DIR")
+        return os.path.basename(env) if env else None
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply hooks/obsidian_utils.py]: ✅ Addressed in a65ab9b.
**What was done:** `_resolve_project_basename` now normalizes falsy basenames to None on both branches:
- cwd basename: empty result → None (prevents cross-project glob from cwd='/')
- env var: `os.path.basename(env.rstrip('/'))` first (so '/path/to/proj/' yields 'proj' not ''), then empty → None (handles CLAUDE_PROJECT_DIR='/' edge case)

Caller now correctly falls through to layer 4 (strict exactly-one bootstrap scan) instead of an unscoped JSONL glob. **Regression tests:** `test_resolve_project_basename_normalizes_root_cwd_to_none`, `test_resolve_project_basename_normalizes_env_trailing_slash`, `test_resolve_project_basename_env_root_normalizes_to_none`. 812/811 + 3 pytest pass.

File: hooks/obsidian_utils.py [hunk 4]
```diff
@@ -469,100 +552,90 @@ def _get_session_id_fast() -> str:
...
-    # during the hook's own invocation (which happens whenever the
-    # hook's downstream code calls a function that triggers this).
-    # Use _safe_mtime so a JSONL deleted or rotated between glob and
-    # stat cannot crash the caller (e.g. load_config).
-    matches = _glob_project_jsonls(safe_project)
-    entries = [(_safe_mtime(p), p) for p in matches]
+    safe_cached = _glob.escape(cached_sid)
+    cached_matches = _glob_project_jsonls(safe_project, f"{safe_cached}.jsonl")
+    if not cached_matches:
+        return None
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply hooks/obsidian_utils.py:592]: ✅ Addressed in a65ab9b.
**What was done:** `_try_bootstrap_fast_path()` now applies `_SID_FILENAME_SAFE.fullmatch()` to `cached_sid` before trusting it — symmetric to the R1 validation added to `_recent_bootstrap_sid()`. Both helpers return SIDs that flow into cache path composition, so both need the same guard. Treats invalid content as a cache miss (returns None → caller falls through to slow path). **Regression test:** `test_try_bootstrap_fast_path_rejects_unsafe_cached_sid` seeds a path-traversal value into the bootstrap and asserts the fast path rejects it.

File: hooks/obsidian_utils.py [hunk 5]
```diff
@@ -277,6 +277,87 @@ def _safe_getcwd() -> str:
...
+
+    Resolution order:
+      1. os.getcwd() — normal case
+      2. CLAUDE_PROJECT_DIR env var — when cwd is gone (worktree deleted
+         mid-session via `gh pr merge --delete-branch`); see issue #105
+      3. None — caller should treat as 'cannot determine project' and
+         fall through to project-agnostic fallbacks
+
+    Never raises. Mirrors the env-fallback contract that obsidian_session_hint.py
+    line 117 already uses for hook input (cwd ← hook payload, getcwd second).
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply hooks/obsidian_utils.py]: ✅ Addressed in f88f454.
**What was done:** Replaced the misleading reference to `obsidian_session_hint.py:117` (which uses `dict.get(key, default)` with eager default evaluation, making the original 'lazy fallback contract' claim inaccurate) with text that describes only this function's own lazy fallback order. Adopted your suggested phrasing.

File: tests/test_get_session_context.py [hunk 6]
```diff
@@ -632,3 +632,275 @@ def test_is_resumed_session_uses_provided_cwd_over_getcwd(tmp_path, monkeypatch)
...
+    bdir.chmod(0o700)
+    # Write an unsafe (path-traversal) value into the bootstrap
+    (bdir / f"sid-{project}").write_text("../../../tmp/escape")
+    monkeypatch.setattr(obsidian_utils, "_BOOTSTRAP_PREFIX", str(bdir) + "/sid-")
+
+    # Seed a JSONL named with the unsafe value too — without validation,
+    # the fast path would otherwise return the unsafe SID.
+    cc_dir = isolated_home / ".claude" / "projects" / f"-Users-test-{project}"
+    cc_dir.mkdir(parents=True, exist_ok=True)
+    (cc_dir / "../../../tmp/escape.jsonl").parent.mkdir(parents=True, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply tests/test_get_session_context.py]: ✅ Addressed in f88f454.
**What was done:** Removed the cc_dir traversal-path mkdir entirely, per your suggestion. The validation guard rejects content before composing JSONL paths, so no JSONL setup is needed. Test now correctly asserts only what its name claims — fast-path rejection of unsafe bootstrap content. Eliminates the accidental fixture-HOME pollution from the `../../../tmp/escape` resolve.

File: scripts/dev-test/test-issue-105-manual.sh [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+    rm -f "$BDIR"/sid-* "$BDIR"/.ob-sid-*.tmp
+    echo -n "test-sid-e1" > "$BDIR/sid-projE1"
+    (
+        TMP=$(mktemp -d -t ob-e1.XXXXXX)
+        cd "$TMP"
+        rmdir "$TMP"
+        unset CLAUDE_PROJECT_DIR
+        # Exact prelude from insight-saver SKILL.md scripts
+        cd "$(git rev-parse --show-toplevel 2>/dev/null || pwd 2>/dev/null || echo /)" 2>/dev/null
+        HOME="$FIXTURE/home" PYTHONPATH="$HOOK_DIR" python3 -c \
```
[the remark this answers is not in the document]
Comment (abhattacherjee) [reply scripts/dev-test/test-issue-105-manual.sh:312]: ✅ Addressed in f88f454.
**What was done:** Updated the comment to document the `|| echo /` defensive variant and explain why the fixture deviates: under `set -euo pipefail` + wedged cwd, the SKILL.md form `|| pwd` would also fail and the `cd` would no-op — meaning python3 wouldn't actually launch under the wedged cwd we're trying to test. The `|| echo /` lets cd land at `/` so python3 launches and exercises the resolver chain. Comment now matches behavior.
